sdk/monitor/azure-monitor-ingestion/src/main/java/com/azure/monitor/ingestion/LogsIngestionAsyncClient.java

```diff
@@ -0,0 +1,268 @@
+    }
+
+    /**
+     * Uploads logs to Azure Monitor with specified data collection rule id and stream name. The input logs may be
+     * too large to be sent as a single request to the Azure Monitor service. In such cases, this method will split
+     * the input logs into multiple smaller requests before sending to the service.
+     * @param dataCollectionRuleId the data collection rule id that is configured to collect and transform the logs.
+     * @param streamName the stream name configured in data collection rule that matches defines the structure of the
+     * logs sent in this request.
+     * @param logs the collection of logs to be uploaded.
+     * @param options the options to configure the upload request.
+     * @return the result of the logs upload request.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<UploadLogsResult> upload(String dataCollectionRuleId, String streamName,
+                                         List<Object> logs, UploadLogsOptions options) {
+        return withContext(context -> upload(dataCollectionRuleId, streamName, logs, options, context));
+    }
+
+    Mono<UploadLogsResult> upload(String dataCollectionRuleId, String streamName,
+                                  List<Object> logs, UploadLogsOptions options,
+                                  Context context) {
+        return Mono.defer(() -> splitAndUpload(dataCollectionRuleId, streamName, logs, options, context));
+    }
+
+    private Mono<UploadLogsResult> splitAndUpload(String dataCollectionRuleId, String streamName, List<Object> logs, UploadLogsOptions options, Context context) {
+        try {
+            Objects.requireNonNull(dataCollectionRuleId, "'dataCollectionRuleId' cannot be null.");
+            Objects.requireNonNull(dataCollectionRuleId, "'streamName' cannot be null.");
+            Objects.requireNonNull(dataCollectionRuleId, "'logs' cannot be null.");
+
+            if (logs.isEmpty()) {
+                throw LOGGER.logExceptionAsError(new IllegalArgumentException("'logs' cannot be empty."));
+            }
+
+            ObjectSerializer serializer = DEFAULT_SERIALIZER;
+
+            // set concurrency to 1 as default
+            int concurrency = 1;
+            if (options != null) {
+                if (options.getObjectSerializer() != null) {
+                    serializer = options.getObjectSerializer();
+                }
+                if (options.getMaxConcurrency() != null) {
+                    concurrency = options.getMaxConcurrency();
+                }
+            }
+
+            List<List<Object>> logBatches = new ArrayList<>();
+            // TODO (srnagar): can improve memory usage by creating these request payloads right before sending the
+            //  request if the allowed concurrency is lower than the number of requests.
+            List<byte[]> requests = createGzipRequests(logs, serializer, logBatches);
```

Comment:
Another change I'd like to see is making `List<List<Object>> logBatches` and `List<byte[]> requests` into a tuple or simple holder object. Right now we're handling to lists with expected index matching, but what if there ends up being a race condition in `map` where `logBatchesIterator` is being used and it misaligns the two list indexes where

request[0] -> logBatches[1]

and

request[1] -> logBatches[0]